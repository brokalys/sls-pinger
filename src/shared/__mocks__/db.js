export const query = jest.fn();
export const getPingersByFrequency = jest.fn().mockReturnValue([]);
export const getPropertyQueueForPingers = jest.fn().mockReturnValue([]);
export const lockPropertyQueueItems = jest.fn();
export const deletePropertyQueueItems = jest.fn();
export const getEmailsThatShouldBeLimitLocked = jest.fn().mockReturnValue([]);
export const getEmailsWithLimitLockerNotification = jest
  .fn()
  .mockReturnValue([]);
export const limitLockPingerEmails = jest.fn();
export const createPingerStatsEntry = jest.fn();
export const logPingerAttempt = jest.fn();
export const updatePingerAttemptTimestamp = jest.fn();
export const getAvailablePingers = jest.fn().mockReturnValue([]);
export const queuePingerForSummaryEmail = jest.fn();
export const getPropertyStats = jest.fn().mockReturnValue([]);
