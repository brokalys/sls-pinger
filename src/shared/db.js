import serverlessMysql from 'serverless-mysql';
import moment from 'moment';

const connection = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
    timezone: 'Z',
    typeCast: true,
  },
});

export function getPingersByType(type) {
  return connection.query({
    sql: `
      SELECT *
      FROM pinger_emails
      WHERE unsubscribed_at IS NULL
        AND (limit_reached_at IS NULL OR limit_reached_at < ? OR is_premium = true)
        AND type = ?
   `,
    values: [moment.utc().startOf('month').toDate(), type],
    typeCast(field, next) {
      if (field.type === 'TINY' && field.length === 1) {
        return field.string() === '1';
      }

      if (field.name === 'categories' || field.name === 'types') {
        return JSON.parse(field.string());
      }

      return next();
    },
  });
}

export function getPropertyQueueForPingers(pingerIds) {
  return connection.query({
    sql: `
      SELECT *
      FROM pinger_queue
      WHERE pinger_id IN (?)
        AND locked = 0
   `,
    values: [pingerIds],
  });
}

export function lockPropertyQueueItems(itemIds) {
  return connection.query({
    sql: `
      UPDATE pinger_queue
      SET locked = 1
      WHERE id IN (?)
   `,
    values: [itemIds],
  });
}

export function deletePropertyQueueItems(itemIds) {
  return connection.query({
    sql: `
      DELETE FROM pinger_queue
      WHERE id IN (?)
   `,
    values: [itemIds],
  });
}
