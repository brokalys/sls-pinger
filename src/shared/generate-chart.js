const vega = require('vega');
const chartBase = require('./data/base-chart.json');
const s3 = require('./s3');

module.exports = async function generateChart(fileName, values, maxDate) {
  chartBase.data[0].values = values;
  chartBase.scales[0].domainMax = maxDate;

  const view = new vega.View(vega.parse(chartBase), {
    renderer: 'none',
  });

  const svgStr = await view.toSVG();
  const bucketName = `${process.env.STAGE}-pinger-charts`;

  await s3
    .putObject({
      Bucket: bucketName,
      Key: fileName,
      Body: svgStr,
      ContentType: 'image/svg+xml',
    })
    .promise();

  return `https://${bucketName}.s3.${process.env.AWS_REGION}.amazonaws.com/${fileName}`;
};
