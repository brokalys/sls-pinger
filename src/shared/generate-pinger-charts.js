const moment = require('moment');
const db = require('./db');
const generateChart = require('./generate-chart');

module.exports = async function generatePingerCharts(pingers) {
  const pingerIds = pingers.map(({ id }) => id);
  const pingerHashMap = pingers.reduce(
    (carry, pinger) => ({ ...carry, [pinger.id]: pinger }),
    {},
  );
  const stats = await db.getPropertyStats(pingerIds);

  const state = {
    ids: [],
    entities: {},
  };

  stats.forEach((row) => {
    const createdAt = moment(row.created_at).format('YYYY-MM-DD');

    if (!state.entities[row.pinger_id]) {
      state.entities[row.pinger_id] = {
        values: [],
        minDate: null,
        maxDate: null,
        uniqueDates: [],
        frequency: pingerHashMap[row.pinger_id].frequency,
      };
      state.ids.push(row.pinger_id);
    }

    if (!state.entities[row.pinger_id].uniqueDates.includes(createdAt)) {
      state.entities[row.pinger_id].uniqueDates.push(createdAt);
    }

    if (
      !state.entities[row.pinger_id].minDate ||
      state.entities[row.pinger_id].minDate > row.created_at
    ) {
      state.entities[row.pinger_id].minDate = row.created_at;
    }

    if (
      !state.entities[row.pinger_id].maxDate ||
      state.entities[row.pinger_id].maxDate < row.created_at
    ) {
      state.entities[row.pinger_id].maxDate = row.created_at;
    }

    row.data.forEach((price) => {
      state.entities[row.pinger_id].values.push({
        x: createdAt,
        y: price,
      });
    });
  });

  const idsWithCharts = state.ids.filter(
    (id) => state.entities[id].uniqueDates.length >= 2,
  );

  const urls = await Promise.all(
    idsWithCharts.map((id) => {
      const idHash = pingerHashMap[id].id_hash;
      const maxDate = moment(state.entities[id].maxDate);
      const calculatedFutureDate = moment(state.entities[id].minDate).add(
        11,
        mapPingerFrequencyToMomentJs(state.entities[id].frequency),
      );

      return generateChart(
        `${idHash}.svg`,
        state.entities[id].values,
        (maxDate.isAfter(calculatedFutureDate)
          ? maxDate
          : calculatedFutureDate
        ).valueOf(),
      );
    }),
  );

  return urls.reduce(
    (carry, url, index) => ({ ...carry, [idsWithCharts[index]]: url }),
    {},
  );
};

function mapPingerFrequencyToMomentJs(frequency) {
  switch (frequency) {
    case 'daily':
      return 'days';
    case 'weekly':
      return 'weeks';
    case 'monthly':
      return 'months';
  }

  return 'days';
}
