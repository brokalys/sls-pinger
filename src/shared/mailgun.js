import mailgun from 'mailgun-js';

export default mailgun({
  apiKey: process.env.MAILGUN_API_KEY,
  domain: process.env.MAILGUN_DOMAIN,
});
