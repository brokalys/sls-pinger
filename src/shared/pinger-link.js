function createUnsubscribeLink(pinger) {
  return `https://unsubscribe.brokalys.com/?key=${encodeURIComponent(
    pinger.unsubscribe_key,
  )}&id=${encodeURIComponent(pinger.id_hash)}`;
}

function createAllPingersLink(pinger) {
  const id = encodeURIComponent(pinger.id_hash);
  const unsubscribeKey = encodeURIComponent(pinger.unsubscribe_key);

  return `https://pinger.brokalys.com/#/pingers/${id},${unsubscribeKey}`;
}

module.exports = {
  unsubscribeLink: createUnsubscribeLink,
  allPingersLink: createAllPingersLink,
};
