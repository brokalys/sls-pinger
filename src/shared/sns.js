const AWS = require('aws-sdk');

AWS.config.update({ region: process.env.AWS_REGION });

module.exports = new AWS.SNS({ apiVersion: '2010-03-31' });
