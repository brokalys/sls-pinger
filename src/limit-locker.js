import * as db from './shared/db';
import sns from './shared/sns';
import * as utils from './shared/utils';

exports.run = async (event, context) => {
  context.callbackWaitsForEmptyEventLoop = false;

  // Get all emails that should be in the `locked` state
  const lockedEmails = await db.getEmailsThatShouldBeLimitLocked();

  if (!lockedEmails.length) {
    return;
  }

  // Lock all the emails that have not yet been locked
  await db.limitLockPingerEmails(lockedEmails);

  // Emails that have received limit-locker notification email
  const limitNotificationEmails = await db.getEmailsWithLimitLockerNotification(
    lockedEmails,
  );

  // Filter emails that have NOT received the notification
  const emails = lockedEmails.filter(
    (email) => !limitNotificationEmails.includes(email),
  );

  if (!emails.length) {
    return;
  }

  // Send all the emails
  await Promise.all(
    emails.map((email) => {
      sns
        .publish({
          Message: 'email',
          MessageAttributes: {
            to: {
              DataType: 'String',
              StringValue: email,
            },
            subject: {
              DataType: 'String',
              StringValue: 'Brokalys ikmēneša e-pastu limits ir sasniegts',
            },
            template_id: {
              DataType: 'String',
              StringValue: 'limit-notification-email',
            },
          },
          MessageStructure: 'string',
          TargetArn: utils.constructArn(
            context,
            process.env.EMAIL_SNS_TOPIC_NAME,
          ),
        })
        .promise();
    }),
  );
};
