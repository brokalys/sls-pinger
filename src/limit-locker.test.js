import moment from 'moment';
import * as db from './shared/db';
import sns from './shared/sns';
import { run } from './limit-locker';
import {
  createPingerFixture,
  createPropertyQueueItemFixture,
} from '__fixtures__';

jest.mock('moment');
jest.mock('./shared/db');
jest.mock('./shared/sns');

const TEST_EMAIL_1 = 'test+1@brokalys.com';
const TEST_EMAIL_2 = 'test+2@brokalys.com';

describe('limit-locker', () => {
  let isBefore;

  // TODO: remove the date hack after February
  beforeEach(() => {
    isBefore = jest.fn().mockReturnValue(false);
    moment.mockReturnValue({ isBefore });
  });

  afterEach(jest.clearAllMocks);

  test('sends an email SNS notification', async () => {
    db.getEmailsThatShouldBeLimitLocked.mockReturnValue([
      TEST_EMAIL_1,
      TEST_EMAIL_2,
    ]);
    db.getEmailsWithLimitLockerNotification.mockReturnValue([]);

    await run();

    expect(sns.publish).toBeCalledTimes(2);
    expect(sns.publish.mock.calls[0]).toMatchSnapshot();
  });

  test('does not send a email SNS notification for the customer that has already received one', async () => {
    db.getEmailsThatShouldBeLimitLocked.mockReturnValue([
      TEST_EMAIL_1,
      TEST_EMAIL_2,
    ]);
    db.getEmailsWithLimitLockerNotification.mockReturnValue([TEST_EMAIL_1]);

    await run();

    expect(sns.publish).toBeCalledTimes(1);
    expect(sns.publish.mock.calls[0]).toMatchSnapshot();
  });

  test('stops early if no emails should be limit-locked', async () => {
    db.getEmailsThatShouldBeLimitLocked.mockReturnValue([]);

    await run();

    expect(db.limitLockPingerEmails).not.toBeCalled();
    expect(sns.publish).not.toBeCalled();
  });

  test('updates the limit-locked timestamp for locked emails', async () => {
    db.getEmailsThatShouldBeLimitLocked.mockReturnValue([TEST_EMAIL_1]);

    await run();

    expect(db.limitLockPingerEmails).toBeCalledWith([TEST_EMAIL_1]);
  });

  // TODO: remove the date hack after February
  test('does not send a email SNS notification if it is not yet February, but still locks', async () => {
    isBefore.mockReturnValue(true);
    db.getEmailsThatShouldBeLimitLocked.mockReturnValue([TEST_EMAIL_1]);
    db.getEmailsWithLimitLockerNotification.mockReturnValue([]);

    await run();

    expect(db.limitLockPingerEmails).toBeCalledWith([TEST_EMAIL_1]);
    expect(sns.publish).not.toBeCalled();
  });
});
