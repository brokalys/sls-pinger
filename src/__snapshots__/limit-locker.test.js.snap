// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`limit-locker does not send a email SNS notification for the customer that has already received one 1`] = `
Array [
  Object {
    "Message": "email",
    "MessageAttributes": Object {
      "subject": Object {
        "DataType": "String",
        "StringValue": "Brokalys ikmēneša e-pastu limits ir sasniegts",
      },
      "template_id": Object {
        "DataType": "String",
        "StringValue": "limit-notification-email",
      },
      "to": Object {
        "DataType": "String",
        "StringValue": "test+2@brokalys.com",
      },
    },
    "MessageStructure": "string",
    "TargetArn": "arn:aws:sns:undefined:173751334418:email-dev",
  },
]
`;

exports[`limit-locker sends an email SNS notification 1`] = `
Array [
  Object {
    "Message": "email",
    "MessageAttributes": Object {
      "subject": Object {
        "DataType": "String",
        "StringValue": "Brokalys ikmēneša e-pastu limits ir sasniegts",
      },
      "template_id": Object {
        "DataType": "String",
        "StringValue": "limit-notification-email",
      },
      "to": Object {
        "DataType": "String",
        "StringValue": "test+1@brokalys.com",
      },
    },
    "MessageStructure": "string",
    "TargetArn": "arn:aws:sns:undefined:173751334418:email-dev",
  },
]
`;
